cobalt/dom/url_media_source.cc

```diff
@@ -32,25 +35,63 @@ void RegisterMediaSourceObjectURL(
     script::EnvironmentSettings* environment_settings,
     const std::string& blob_url,
     const scoped_refptr<dom::MediaSource>& media_source) {
-  dom::DOMSettings* dom_settings =
-      base::polymorphic_downcast<dom::DOMSettings*>(environment_settings);
-  DCHECK(dom_settings);
-  DCHECK(dom_settings->media_source_registry());
+  web::EnvironmentSettings* web_settings =
+      base::polymorphic_downcast<web::EnvironmentSettings*>(
+          environment_settings);
+  DCHECK(web_settings);
+  DCHECK(web_settings->context());
+  web::WindowOrWorkerGlobalScope* global_scope =
+      web_settings->context()->GetWindowOrWorkerGlobalScope();
 
-  scoped_refptr<MediaSourceAttachment> attachment =
-      base::MakeRefCounted<SameThreadMediaSourceAttachment>(media_source);
+  if (global_scope->IsWorker()) {
+    worker::WorkerSettings* worker_settings =
+        base::polymorphic_downcast<worker::WorkerSettings*>(
+            environment_settings);
+    DCHECK(worker_settings);
+    DCHECK(worker_settings->media_source_registry());
 
-  dom_settings->media_source_registry()->Register(blob_url, attachment);
+    scoped_refptr<MediaSourceAttachment> attachment =
+        base::MakeRefCounted<CrossThreadMediaSourceAttachment>(media_source);
+
+    worker_settings->media_source_registry()->Register(blob_url, attachment);
+  } else {
+    dom::DOMSettings* dom_settings =
+        base::polymorphic_downcast<dom::DOMSettings*>(environment_settings);
+    DCHECK(dom_settings);
+    DCHECK(dom_settings->media_source_registry());
+
+    scoped_refptr<MediaSourceAttachment> attachment =
+        base::MakeRefCounted<SameThreadMediaSourceAttachment>(media_source);
+
+    dom_settings->media_source_registry()->Register(blob_url, attachment);
+  }
 }
 
 // extern
 bool UnregisterMediaSourceObjectURL(
     script::EnvironmentSettings* environment_settings, const std::string& url) {
-  dom::DOMSettings* dom_settings =
-      base::polymorphic_downcast<dom::DOMSettings*>(environment_settings);
-  DCHECK(dom_settings);
-  DCHECK(dom_settings->media_source_registry());
-  return dom_settings->media_source_registry()->Unregister(url);
+  web::EnvironmentSettings* web_environment_settings =
+      base::polymorphic_downcast<web::EnvironmentSettings*>(
+          environment_settings);
+  DCHECK(web_environment_settings);
+  DCHECK(web_environment_settings->context());
+  web::WindowOrWorkerGlobalScope* global_scope =
+      web_environment_settings->context()->GetWindowOrWorkerGlobalScope();
+
+  if (global_scope->IsWorker()) {
```

Comment:
Same for here. Is this `IsDedicatedWorker` or `IsWorker`?

Reply:
This should work regardless of the worker type, since it's checking only the WorkerSettings.